Epinio cli should work without kube access

There are some commands that will always need access to the kube api. E.g.
- epinio install
- epinio config update-credentials

These are considered "admin" commands. Regular epinio users should not have to run those commands. This means, all the commands that are not "admin" commands should be possible to run with no access to the kubernetes cluster. That is, with no kubeconfig in place.

Right now Epinio is still talking to the kube api for various other reasons. E.g. for shell autocompletion (find available services/apps/etc) or in order to discover the Epinio API by looking up for the ingress.

## Acceptance Criteria

- All the code that fetches information from the cluster by talking to kube is migrated to API endpoints
- No API endpoint autodiscovery happens. There should be a command to target an instance of Epinio (e.g. `epinio login`). This command should update the epinio config file if run successfully. This command is an alternative to running `epinio config update-credentials` as an admin. It can be used when no kubernetes cluster access is possible.

Removing the `Cluster` field:

> internal/cli/clients/client.go:84:3: unknown field 'Cluster' in struct literal of type EpinioClient
internal/cli/clients/client.go:809:15: c.Cluster undefined (type *EpinioClient has no field or method Cluster)
internal/cli/clients/client.go:810:23: c.Cluster undefined (type *EpinioClient has no field or method Cluster)
internal/cli/clients/client.go:1560:38: c.Cluster undefined (type *EpinioClient has no field or method Cluster)
internal/cli/clients/client.go:1566:33: c.Cluster undefined (type *EpinioClient has no field or method Cluster)
internal/cli/clients/push.go:140:34: c.Cluster undefined (type *EpinioClient has no field or method Cluster)
internal/cli/clients/push.go:175:10: c.Cluster undefined (type *EpinioClient has no field or method Cluster)

Lets work through that.

--- a/acceptance/api/v1/applications_test.go
+++ b/acceptance/api/v1/applications_test.go
@@ -141,7 +141,6 @@ var _ = Describe("Apps API Application Endpoints", func() {
 				Revision: uploadResponse.Git.Revision,
 				URL:      uploadResponse.Git.URL,
 			},
-			Route:        appName + ".omg.howdoi.website",
 			BuilderImage: "paketobuildpacks/builder:full",
 		}
 		b, err := json.Marshal(request)
@@ -462,7 +461,6 @@ var _ = Describe("Apps API Application Endpoints", func() {
 						Stage: models.StageRef{
 							ID: stageResponse.Stage.ID,
 						},
-						Route: appName + ".omg.howdoi.website",
 						Git: &models.GitRef{
 							Revision: respObj.Git.Revision,
 							URL:      respObj.Git.URL,
@@ -485,6 +483,11 @@ var _ = Describe("Apps API Application Endpoints", func() {
 					Expect(err).ToNot(HaveOccurred())
 					Expect(response.StatusCode).To(Equal(http.StatusOK), string(bodyBytes))
 
+					deploy := &models.DeployResponse{}
+					err = json.Unmarshal(bodyBytes, deploy)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(deploy.Route).To(MatchRegexp(appName + `.*\.omg\.howdoi\.website`))
+
 					Eventually(func() string {
 						return appStatus(org, appName)
 					}, "5m").Should(Equal("1/1"))
@@ -500,7 +503,6 @@ var _ = Describe("Apps API Application Endpoints", func() {
 						Org:  org,
 					},
 					Instances: &one,
-					Route:     appName + ".omg.howdoi.website",
 					ImageURL:  "splatform/sample-app",
 				}
 
@@ -523,6 +525,11 @@ var _ = Describe("Apps API Application Endpoints", func() {
 					bodyBytes, err := ioutil.ReadAll(response.Body)
 					Expect(err).ToNot(HaveOccurred())
 					Expect(response.StatusCode).To(Equal(http.StatusOK), string(bodyBytes))
+
+					deploy := &models.DeployResponse{}
+					err = json.Unmarshal(bodyBytes, deploy)
+					Expect(err).NotTo(HaveOccurred())
+					Expect(deploy.Route).To(MatchRegexp(appName + `.*\.omg\.howdoi\.website`))
 
 					Eventually(func() string {
 						return appStatus(org, appName)

--- a/acceptance/api/v1/applications_test.go
+++ b/acceptance/api/v1/applications_test.go
@@ -449,9 +449,10 @@ var _ = Describe("Apps API Application Endpoints", func() {
 					err = json.Unmarshal(bodyBytes, &respObj)
 					Expect(err).ToNot(HaveOccurred())
 
-					By("creating staging resource first")
+					By("staging first")
 					stageResponse := stageApplication(appName, org, respObj)
 
+					By("deploying the staged resource")
 					request = models.DeployRequest{
 						App: models.AppRef{
 							Name: appName,
@@ -488,6 +489,17 @@ var _ = Describe("Apps API Application Endpoints", func() {
 					Expect(err).NotTo(HaveOccurred())
 					Expect(deploy.Route).To(MatchRegexp(appName + `.*\.omg\.howdoi\.website`))
 
+					By("waiting for the deployment to complete")
+
+					url = serverURL + "/" + v1.Routes.Path("AppRunning", org, appName)
+
+					response, err = env.Curl("GET", url, strings.NewReader(body))
+					Expect(err).ToNot(HaveOccurred())
+					Expect(response).ToNot(BeNil())
+					defer response.Body.Close()
+
+					By("confirming at highlevel")
+					// Highlevel check and confirmation
 					Eventually(func() string {
 						return appStatus(org, appName)
 					}, "5m").Should(Equal("1/1"))

--- a/helpers/kubernetes/cluster.go
+++ b/helpers/kubernetes/cluster.go
@@ -357,8 +357,10 @@ func (c *Cluster) IsDeploymentCompleted(ctx context.Context, deploymentName, nam
 }
 
 func (c *Cluster) WaitForDeploymentCompleted(ctx context.Context, ui *termui.UI, namespace, deploymentName string, timeout time.Duration) error {
-	s := ui.Progressf("Waiting for deployment %s in %s to be ready", deploymentName, namespace)
-	defer s.Stop()
+	if ui != nil {
+		s := ui.Progressf("Waiting for deployment %s in %s to be ready", deploymentName, namespace)
+		defer s.Stop()
+	}
 
 	return wait.PollImmediate(time.Second, timeout, c.IsDeploymentCompleted(ctx, deploymentName, namespace))
 }

--- a/internal/api/v1/applications.go
+++ b/internal/api/v1/applications.go
@@ -168,6 +168,62 @@ func (hc ApplicationsController) Show(w http.ResponseWriter, r *http.Request) AP
 	return nil
 }
 
+func (hc ApplicationsController) ServiceApps(w http.ResponseWriter, r *http.Request) APIErrors {
+	ctx := r.Context()
+	params := httprouter.ParamsFromContext(ctx)
+	org := params.ByName("org")
+
+	cluster, err := kubernetes.GetCluster(ctx)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	exists, err := organizations.Exists(ctx, cluster, org)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	if !exists {
+		return OrgIsNotKnown(org)
+	}
+
+	var appsOf = map[string]models.AppList{}
+
+	apps, err := application.List(ctx, cluster, org)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	for _, app := range apps {
+		w := application.NewWorkload(cluster, app.AppRef())
+		bound, err := w.Services(ctx)
+		if err != nil {
+			return InternalError(err)
+		}
+		for _, bonded := range bound {
+			bname := bonded.Name()
+			if theapps, found := appsOf[bname]; found {
+				appsOf[bname] = append(theapps, app)
+			} else {
+				appsOf[bname] = models.AppList{app}
+			}
+		}
+	}
+
+	js, err := json.Marshal(appsOf)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	_, err = w.Write(js)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	return nil
+}
+
 func (hc ApplicationsController) Update(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)

--- a/internal/api/v1/applications.go
+++ b/internal/api/v1/applications.go
@@ -14,6 +14,7 @@ import (
 	"github.com/epinio/epinio/internal/api/v1/models"
 	"github.com/epinio/epinio/internal/application"
 	"github.com/epinio/epinio/internal/cli/clients/gitea"
+	"github.com/epinio/epinio/internal/duration"
 	"github.com/epinio/epinio/internal/organizations"
 	"github.com/gorilla/websocket"
 	"github.com/julienschmidt/httprouter"
@@ -233,6 +234,54 @@ func (hc ApplicationsController) Update(w http.ResponseWriter, r *http.Request) 
 
 	return nil
 }
+
+func (hc ApplicationsController) Running(w http.ResponseWriter, r *http.Request) APIErrors {
+	ctx := r.Context()
+	params := httprouter.ParamsFromContext(ctx)
+	org := params.ByName("org")
+	appName := params.ByName("app")
+
+	cluster, err := kubernetes.GetCluster(ctx)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	exists, err := organizations.Exists(ctx, cluster, org)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	if !exists {
+		return OrgIsNotKnown(org)
+	}
+
+	exists, err = application.Exists(ctx, cluster, models.NewAppRef(appName, org))
+	if err != nil {
+		return InternalError(err)
+	}
+
+	if !exists {
+		return AppIsNotKnown(appName)
+	}
+
+	app, err := application.Lookup(ctx, cluster, org, appName)
+	if err != nil {
+		return InternalError(err)
+	}
+	if app == nil {
+		// While app exists it has no workload
+		return NewAPIError("No status available for application without workload", "", http.StatusBadRequest)
+	}
+
+	err = cluster.WaitForDeploymentCompleted(
+		ctx, nil, org, appName, duration.ToAppBuilt())
+	if err != nil {
+		return InternalError(err)
+	}
+
+	return nil
+}
+
 func (hc ApplicationsController) Logs(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)

--- a/internal/api/v1/applications.go
+++ b/internal/api/v1/applications.go
@@ -27,6 +27,7 @@ type ApplicationsController struct {
 	conn *websocket.Conn
 }
 
+// Create creates a new and empty application. I.e. without a workload.
 func (hc ApplicationsController) Create(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -74,6 +75,7 @@ func (hc ApplicationsController) Create(w http.ResponseWriter, r *http.Request) 
 	return nil
 }
 
+// Index lists all the known applications, with and without workload.
 func (hc ApplicationsController) Index(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -112,6 +114,7 @@ func (hc ApplicationsController) Index(w http.ResponseWriter, r *http.Request) A
 	return nil
 }
 
+// Show returns the details of the specified application.
 func (hc ApplicationsController) Show(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -168,6 +171,9 @@ func (hc ApplicationsController) Show(w http.ResponseWriter, r *http.Request) AP
 	return nil
 }
 
+// ServiceApps returns a map from services to the apps they are bound to, in the specified org.
+// Internally it asks each app in the org for its bound services and then inverts that map to
+// get the desired result.
 func (hc ApplicationsController) ServiceApps(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -224,6 +230,8 @@ func (hc ApplicationsController) ServiceApps(w http.ResponseWriter, r *http.Requ
 	return nil
 }
 
+// Update modifies the specified application. currently this is only
+// the number of instances to run.
 func (hc ApplicationsController) Update(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -291,6 +299,11 @@ func (hc ApplicationsController) Update(w http.ResponseWriter, r *http.Request) 
 	return nil
 }
 
+// Running waits for the specified application to be running (i.e. its
+// deployment to be complete), before it returns. An exception is if
+// the application does not become running without `duration.ToAppBuilt()`
+// (default: 10 minutes). In that case it returns with an error after
+// that time.
 func (hc ApplicationsController) Running(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -338,6 +351,8 @@ func (hc ApplicationsController) Running(w http.ResponseWriter, r *http.Request)
 	return nil
 }
 
+// Logs arranges for the logs of the specified application to be
+// streamed over a websocket.
 func (hc ApplicationsController) Logs(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)
@@ -497,6 +512,7 @@ func (hc ApplicationsController) streamPodLogs(ctx context.Context, orgName, app
 	return hc.conn.Close()
 }
 
+// Delete removes the application
 func (hc ApplicationsController) Delete(w http.ResponseWriter, r *http.Request) APIErrors {
 	ctx := r.Context()
 	params := httprouter.ParamsFromContext(ctx)

--- a/internal/api/v1/deploy.go
+++ b/internal/api/v1/deploy.go
@@ -8,7 +8,6 @@ import (
 	"net/http"
 
 	"github.com/epinio/epinio/internal/application"
-	"github.com/epinio/epinio/internal/domain"
 	"github.com/epinio/epinio/internal/names"
 	"github.com/julienschmidt/httprouter"
 	appsv1 "k8s.io/api/apps/v1"
@@ -22,6 +21,7 @@ import (
 	"github.com/epinio/epinio/helpers/kubernetes"
 	"github.com/epinio/epinio/helpers/tracelog"
 	"github.com/epinio/epinio/internal/api/v1/models"
+	"github.com/epinio/epinio/internal/domain"
 )
 
 const (
@@ -31,10 +31,8 @@ const (
 type deployParam struct {
 	models.AppRef
 	Git         *models.GitRef
-	Route       string
 	ImageURL    string
 	Instances   int32
-	Domain      string
 	Stage       models.StageRef
 	Owner       metav1.OwnerReference
 	Environment models.EnvVariableList
@@ -102,7 +100,7 @@ func (hc ApplicationsController) Deploy(w http.ResponseWriter, r *http.Request) 
 		}
 	}
 
-	mainDomain, err := domain.MainDomain(ctx)
+	route, err := appDefaultRoute(ctx, req.App.Name)
 	if err != nil {
 		return InternalError(err)
 	}
@@ -116,11 +114,9 @@ func (hc ApplicationsController) Deploy(w http.ResponseWriter, r *http.Request) 
 	deployParams := deployParam{
 		AppRef:      req.App,
 		Git:         req.Git,
-		Route:       req.Route,
 		Owner:       owner,
 		Environment: environment,
 		Instances:   instances,
-		Domain:      mainDomain,
 		ImageURL:    req.ImageURL,
 	}
 
@@ -167,9 +163,9 @@ func (hc ApplicationsController) Deploy(w http.ResponseWriter, r *http.Request) 
 		}
 	}
 
-	log.Info("deploying app ingress", "org", org, "app", req.App, "", req.Route)
+	log.Info("deploying app ingress", "org", org, "app", req.App, "", route)
 
-	ing, err := newAppIngress(req.App, req.Route)
+	ing, err := newAppIngress(req.App, route)
 	if err != nil {
 		return InternalError(err)
 	}
@@ -192,6 +188,14 @@ func (hc ApplicationsController) Deploy(w http.ResponseWriter, r *http.Request) 
 		if err := application.Unstage(ctx, cluster, req.App, req.Stage.ID); err != nil {
 			return InternalError(err)
 		}
+	}
+
+	resp := models.DeployResponse{
+		Route: route,
+	}
+	err = jsonResponse(w, resp)
+	if err != nil {
+		return InternalError(err)
 	}
 
 	return nil
@@ -360,4 +364,12 @@ func existingReplica(ctx context.Context, client *k8s.Clientset, app models.AppR
 		return 0, err
 	}
 	return *result.Spec.Replicas, nil
+}
+
+func appDefaultRoute(ctx context.Context, name string) (string, error) {
+	mainDomain, err := domain.MainDomain(ctx)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s.%s", name, mainDomain), nil
 }

--- a/internal/api/v1/deploy.go
+++ b/internal/api/v1/deploy.go
@@ -100,7 +100,7 @@ func (hc ApplicationsController) Deploy(w http.ResponseWriter, r *http.Request) 
 		}
 	}
 
-	route, err := appDefaultRoute(ctx, req.App.Name)
+	route, err := domain.AppDefaultRoute(ctx, req.App.Name)
 	if err != nil {
 		return InternalError(err)
 	}
@@ -364,12 +364,4 @@ func existingReplica(ctx context.Context, client *k8s.Clientset, app models.AppR
 		return 0, err
 	}
 	return *result.Spec.Replicas, nil
-}
-
-func appDefaultRoute(ctx context.Context, name string) (string, error) {
-	mainDomain, err := domain.MainDomain(ctx)
-	if err != nil {
-		return "", err
-	}
-	return fmt.Sprintf("%s.%s", name, mainDomain), nil
 }

--- a/internal/api/v1/info.go
+++ b/internal/api/v1/info.go
@@ -4,6 +4,7 @@ import (
 	"encoding/json"
 	"net/http"
 
+	"github.com/epinio/epinio/helpers/kubernetes"
 	"github.com/epinio/epinio/internal/version"
 )
 
@@ -11,11 +12,30 @@ type InfoController struct {
 }
 
 func (hc InfoController) Info(w http.ResponseWriter, r *http.Request) APIErrors {
-	info := struct {
-		Version string
-	}{
-		Version: version.Version,
+	ctx := r.Context()
+
+	cluster, err := kubernetes.GetCluster(ctx)
+	if err != nil {
+		return InternalError(err)
 	}
+
+	kubeVersion, err := cluster.GetVersion()
+	if err != nil {
+		return InternalError(err)
+	}
+
+	platform := cluster.GetPlatform()
+
+	info := struct {
+		Version     string
+		KubeVersion string
+		Platform    string
+	}{
+		Version:     version.Version,
+		Platform:    platform.String(),
+		KubeVersion: kubeVersion,
+	}
+
 	js, err := json.Marshal(info)
 	if err != nil {
 		return InternalError(err)

--- a/internal/api/v1/models/models.go
+++ b/internal/api/v1/models/models.go
@@ -57,7 +57,6 @@ type UploadResponse struct {
 type StageRequest struct {
 	App          AppRef  `json:"app,omitempty"`
 	Git          *GitRef `json:"git,omitempty"`
-	Route        string  `json:"route,omitempty"`
 	BuilderImage string  `json:"builderimage,omitempty"`
 }
 
@@ -70,9 +69,12 @@ type DeployRequest struct {
 	App       AppRef   `json:"app,omitempty"`
 	Instances *int32   `json:"instances,omitempty"`
 	Stage     StageRef `json:"stage,omitempty"`
-	Route     string   `json:"route,omitempty"`
 	Git       *GitRef  `json:"git,omitempty"`
 	ImageURL  string   `json:"image,omitempty"`
+}
+
+type DeployResponse struct {
+	Route string `json:"route,omitempty"`
 }
 
 type ApplicationDeleteResponse struct {

--- a/internal/api/v1/router.go
+++ b/internal/api/v1/router.go
@@ -99,6 +99,8 @@ var Routes = routes.NamedRoutes{
 	"OrgDelete": delete("/orgs/:org", errorHandler(OrganizationsController{}.Delete)),
 
 	// List, show, create and delete services, catalog and custom
+	"ServiceApps": get("/orgs/:org/serviceapps", errorHandler(ApplicationsController{}.ServiceApps)),
+	//
 	"Services":            get("/orgs/:org/services", errorHandler(ServicesController{}.Index)),
 	"ServiceShow":         get("/orgs/:org/services/:service", errorHandler(ServicesController{}.Show)),
 	"ServiceCreate":       post("/orgs/:org/services", errorHandler(ServicesController{}.Create)),

--- a/internal/api/v1/router.go
+++ b/internal/api/v1/router.go
@@ -67,16 +67,18 @@ func patch(path string, h http.HandlerFunc) routes.Route {
 var Routes = routes.NamedRoutes{
 	"Info": get("/info", errorHandler(InfoController{}.Info)),
 
-	"Apps":        get("/orgs/:org/applications", errorHandler(ApplicationsController{}.Index)),
-	"AppCreate":   post("/orgs/:org/applications", errorHandler(ApplicationsController{}.Create)),
-	"AppShow":     get("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Show)),
-	"AppLogs":     get("/orgs/:org/applications/:app/logs", ApplicationsController{}.Logs),
-	"StagingLogs": get("/orgs/:org/staging/:stage_id/logs", ApplicationsController{}.Logs),
-	"AppDelete":   delete("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Delete)),
-	"AppUpload":   post("/orgs/:org/applications/:app/store", errorHandler(ApplicationsController{}.Upload)), // See upload.go
-	"AppStage":    post("/orgs/:org/applications/:app/stage", errorHandler(ApplicationsController{}.Stage)),  // See stage.go
-	"AppDeploy":   post("/orgs/:org/applications/:app/deploy", errorHandler(ApplicationsController{}.Deploy)),
-	"AppUpdate":   patch("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Update)),
+	"Apps":            get("/orgs/:org/applications", errorHandler(ApplicationsController{}.Index)),
+	"AppCreate":       post("/orgs/:org/applications", errorHandler(ApplicationsController{}.Create)),
+	"AppShow":         get("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Show)),
+	"AppLogs":         get("/orgs/:org/applications/:app/logs", ApplicationsController{}.Logs),
+	"StagingLogs":     get("/orgs/:org/staging/:stage_id/logs", ApplicationsController{}.Logs),
+	"StagingComplete": get("/orgs/:org/staging/:stage_id/complete", errorHandler(ApplicationsController{}.Staged)), // See stage.go
+	"AppDelete":       delete("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Delete)),
+	"AppUpload":       post("/orgs/:org/applications/:app/store", errorHandler(ApplicationsController{}.Upload)), // See upload.go
+	"AppStage":        post("/orgs/:org/applications/:app/stage", errorHandler(ApplicationsController{}.Stage)),  // See stage.go
+	"AppDeploy":       post("/orgs/:org/applications/:app/deploy", errorHandler(ApplicationsController{}.Deploy)),
+	"AppUpdate":       patch("/orgs/:org/applications/:app", errorHandler(ApplicationsController{}.Update)),
+	"AppRunning":      get("/orgs/:org/applications/:app/running", errorHandler(ApplicationsController{}.Running)),
 
 	// See env.go
 	"EnvList":  get("/orgs/:org/applications/:app/environment", errorHandler(ApplicationsController{}.EnvIndex)),

--- a/internal/api/v1/stage.go
+++ b/internal/api/v1/stage.go
@@ -3,19 +3,23 @@ package v1
 import (
 	"context"
 	"encoding/json"
+	"errors"
 	"fmt"
 	"io/ioutil"
 	"net/http"
+	"time"
 
 	"github.com/julienschmidt/httprouter"
 	"github.com/spf13/viper"
 	v1beta1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
 	"github.com/tektoncd/pipeline/pkg/apis/resource/v1alpha1"
 	"github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
+	tekton "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
 	corev1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/api/resource"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/epinio/epinio/deployments"
 	"github.com/epinio/epinio/helpers/kubernetes"
@@ -25,6 +29,8 @@ import (
 	"github.com/epinio/epinio/internal/application"
 	"github.com/epinio/epinio/internal/auth"
 	"github.com/epinio/epinio/internal/domain"
+	"github.com/epinio/epinio/internal/duration"
+	"github.com/epinio/epinio/internal/organizations"
 )
 
 const (
@@ -215,6 +221,67 @@ func (hc ApplicationsController) Stage(w http.ResponseWriter, r *http.Request) A
 		ImageURL: params.ImageURL(params.RegistryURL),
 	}
 	err = jsonResponse(w, resp)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	return nil
+}
+
+// Staged will wait for the Tekton PipelineRun resource staging the app to complete
+func (hc ApplicationsController) Staged(w http.ResponseWriter, r *http.Request) APIErrors {
+	ctx := r.Context()
+
+	p := httprouter.ParamsFromContext(ctx)
+	org := p.ByName("org")
+	id := p.ByName("stage_id")
+
+	cluster, err := kubernetes.GetCluster(ctx)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	exists, err := organizations.Exists(ctx, cluster, org)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	if !exists {
+		return InternalError(err)
+	}
+
+	cs, err := tekton.NewForConfig(cluster.RestConfig)
+	if err != nil {
+		return InternalError(err)
+	}
+
+	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
+
+	err = wait.PollImmediate(time.Second, duration.ToAppBuilt(),
+		func() (bool, error) {
+			l, err := client.List(ctx, metav1.ListOptions{LabelSelector: models.EpinioStageIDLabel + "=" + id})
+			if err != nil {
+				return false, err
+			}
+			if len(l.Items) == 0 {
+				return false, nil
+			}
+			for _, pr := range l.Items {
+				// any failed conditions, throw an error so we can exit early
+				for _, c := range pr.Status.Conditions {
+					if c.IsFalse() {
+						return false, errors.New(c.Message)
+					}
+				}
+				// it worked
+				if pr.Status.CompletionTime != nil {
+					return true, nil
+				}
+			}
+			// pr exists, but still running
+			return false, nil
+		})
+
 	if err != nil {
 		return InternalError(err)
 	}

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -842,12 +842,16 @@ func (c *EpinioClient) AppsMatching(ctx context.Context, prefix string) []string
 
 	result := []string{}
 
-	// TODO Create and use server endpoints. Maybe use existing
-	// `Index`/Listing endpoint, either with parameter for
-	// matching, or local matching.
+	// Ask for all apps. Filtering is local.
+	// TODO: Create new endpoint (compare `EnvMatch`) and move filtering to the server.
 
-	apps, err := application.List(ctx, c.Cluster, c.Config.Org)
+	jsonResponse, err := c.get(api.Routes.Path("Apps", c.Config.Org))
 	if err != nil {
+		return result
+	}
+
+	var apps models.AppList
+	if err := json.Unmarshal(jsonResponse, &apps); err != nil {
 		return result
 	}
 
@@ -859,6 +863,8 @@ func (c *EpinioClient) AppsMatching(ctx context.Context, prefix string) []string
 			result = append(result, app.Name)
 		}
 	}
+
+	sort.Strings(result)
 
 	return result
 }

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -344,17 +344,15 @@ func (c *EpinioClient) ServicePlanMatching(ctx context.Context, serviceClassName
 
 	result := []string{}
 
-	// TODO Create and use server endpoints. Maybe use existing
-	// `Index`/Listing endpoint, either with parameter for
-	// matching, or local matching.
+	// Ask for all service plans of a service class. Filtering is local.
+	// TODO: Create new endpoint (compare `EnvMatch`) and move filtering to the server.
 
-	serviceClass, err := services.ClassLookup(ctx, c.Cluster, serviceClassName)
+	jsonResponse, err := c.get(api.Routes.Path("ServicePlans", serviceClassName))
 	if err != nil {
 		return result
 	}
-
-	servicePlans, err := serviceClass.ListPlans(ctx)
-	if err != nil {
+	var servicePlans services.ServicePlanList
+	if err := json.Unmarshal(jsonResponse, &servicePlans); err != nil {
 		return result
 	}
 

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -378,13 +378,15 @@ func (c *EpinioClient) ServiceClassMatching(ctx context.Context, prefix string) 
 
 	result := []string{}
 
-	// TODO Create and use server endpoints. Maybe use existing
-	// `Index`/Listing endpoint, either with parameter for
-	// matching, or local matching.
+	// Ask for all service classes. Filtering is local.
+	// TODO: Create new endpoint (compare `EnvMatch`) and move filtering to the server.
 
-	serviceClasses, err := services.ListClasses(ctx, c.Cluster)
+	jsonResponse, err := c.get(api.Routes.Path("ServiceClasses"))
 	if err != nil {
-		details.Info("Error", err)
+		return result
+	}
+	var serviceClasses services.ServiceClassList
+	if err := json.Unmarshal(jsonResponse, &serviceClasses); err != nil {
 		return result
 	}
 

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -1723,6 +1723,10 @@ func uniqueStrings(stringSlice []string) []string {
 }
 
 func getAPI(ctx context.Context, log logr.Logger) (string, string, error) {
+	// This is called only by the admin command `config update`
+	// which has to talk to the cluster to retrieve the
+	// information. This is allowed.
+
 	cluster, err := kubernetes.GetCluster(ctx)
 	if err != nil {
 		return "", "", err
@@ -1740,6 +1744,10 @@ func getAPI(ctx context.Context, log logr.Logger) (string, string, error) {
 
 // TODO: https://github.com/epinio/epinio/issues/667
 func getCredentials(ctx context.Context, log logr.Logger) (string, string, error) {
+	// This is called only by the admin command `config update`
+	// which has to talk to the cluster to retrieve the
+	// information. This is allowed.
+
 	cluster, err := kubernetes.GetCluster(ctx)
 	if err != nil {
 		return "", "", err
@@ -1777,6 +1785,10 @@ func getCredentials(ctx context.Context, log logr.Logger) (string, string, error
 }
 
 func getCerts(ctx context.Context, log logr.Logger) (string, error) {
+	// This is called only by the admin command `config update`
+	// which has to talk to the cluster to retrieve the
+	// information. This is allowed.
+
 	// Save the  CA cert into the config. The regular client
 	// will then extend the Cert pool with the same, so that it
 	// can cerify the server cert.

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -28,7 +28,6 @@ import (
 	"github.com/epinio/epinio/internal/api/v1/models"
 	"github.com/epinio/epinio/internal/cli/config"
 	"github.com/epinio/epinio/internal/cli/logprinter"
-	"github.com/epinio/epinio/internal/domain"
 	"github.com/epinio/epinio/internal/duration"
 	"github.com/epinio/epinio/internal/services"
 
@@ -1345,11 +1344,6 @@ func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
 		return fmt.Errorf("%s: %s", "app name incorrect", strings.Join(errorMsgs, "\n"))
 	}
 
-	route, err := appDefaultRoute(ctx, appRef.Name)
-	if err != nil {
-		return errors.Wrap(err, "unable to determine default app route")
-	}
-
 	c.ui.Normal().Msg("Create the application resource ...")
 
 	request := models.ApplicationCreateRequest{Name: appRef.Name}
@@ -1409,7 +1403,6 @@ func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
 		req := models.StageRequest{
 			App:          appRef,
 			Git:          gitRef,
-			Route:        route,
 			BuilderImage: params.BuilderImage,
 		}
 		details.Info("staging code", "Git", gitRef.Revision)
@@ -1431,7 +1424,6 @@ func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
 	deployRequest := models.DeployRequest{
 		App:       appRef,
 		Instances: params.Instances,
-		Route:     route,
 		Git:       gitRef,
 	}
 	// If docker param is specified, then we just take it into ImageURL
@@ -1443,7 +1435,7 @@ func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
 		deployRequest.Stage = models.StageRef{ID: stageID}
 	}
 
-	_, err = c.deployCode(deployRequest)
+	deployResponse, err := c.deployCode(deployRequest)
 	if err != nil {
 		return err
 	}
@@ -1500,19 +1492,11 @@ func (c *EpinioClient) Push(ctx context.Context, params PushParams) error {
 	c.ui.Success().
 		WithStringValue("Name", appRef.Name).
 		WithStringValue("Organization", appRef.Org).
-		WithStringValue("Route", fmt.Sprintf("https://%s", route)).
+		WithStringValue("Route", fmt.Sprintf("https://%s", deployResponse.Route)).
 		WithStringValue("Builder Image", params.BuilderImage).
 		Msg("App is online.")
 
 	return nil
-}
-
-func appDefaultRoute(ctx context.Context, name string) (string, error) {
-	mainDomain, err := domain.MainDomain(ctx)
-	if err != nil {
-		return "", err
-	}
-	return fmt.Sprintf("%s.%s", name, mainDomain), nil
 }
 
 // Target targets an org in gitea

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -26,7 +26,6 @@ import (
 	"github.com/epinio/epinio/helpers/tracelog"
 	api "github.com/epinio/epinio/internal/api/v1"
 	"github.com/epinio/epinio/internal/api/v1/models"
-	"github.com/epinio/epinio/internal/application"
 	"github.com/epinio/epinio/internal/cli/config"
 	"github.com/epinio/epinio/internal/cli/logprinter"
 	"github.com/epinio/epinio/internal/domain"
@@ -1558,27 +1557,14 @@ func (c *EpinioClient) ServicesToApps(ctx context.Context, org string) (map[stri
 	// (inversion of services bound to apps)
 	// Literally query apps in the org for their services and invert.
 
-	var appsOf = map[string]models.AppList{}
-
-	apps, err := application.List(ctx, c.Cluster, c.Config.Org)
+	jsonResponse, err := c.get(api.Routes.Path("ServiceApps", c.Config.Org))
 	if err != nil {
 		return nil, err
 	}
 
-	for _, app := range apps {
-		w := application.NewWorkload(c.Cluster, app.AppRef())
-		bound, err := w.Services(ctx)
-		if err != nil {
-			return nil, err
-		}
-		for _, bonded := range bound {
-			bname := bonded.Name()
-			if theapps, found := appsOf[bname]; found {
-				appsOf[bname] = append(theapps, app)
-			} else {
-				appsOf[bname] = models.AppList{app}
-			}
-		}
+	var appsOf map[string]models.AppList
+	if err := json.Unmarshal(jsonResponse, &appsOf); err != nil {
+		return nil, err
 	}
 
 	return appsOf, nil

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -40,7 +40,6 @@ import (
 // EpinioClient provides functionality for talking to a
 // Epinio installation on Kubernetes
 type EpinioClient struct {
-	Cluster     *kubernetes.Cluster
 	Config      *config.Config
 	Log         logr.Logger
 	ui          *termui.UI
@@ -64,11 +63,6 @@ func NewEpinioClient(ctx context.Context) (*EpinioClient, error) {
 		return nil, err
 	}
 
-	cluster, err := kubernetes.GetCluster(ctx)
-	if err != nil {
-		return nil, err
-	}
-
 	uiUI := termui.NewUI()
 	epClient, err := GetEpinioAPIClient(ctx)
 	if err != nil {
@@ -79,7 +73,6 @@ func NewEpinioClient(ctx context.Context) (*EpinioClient, error) {
 
 	logger := tracelog.NewClientLogger()
 	epinioClient := &EpinioClient{
-		Cluster:     cluster,
 		ui:          uiUI,
 		Config:      configConfig,
 		Log:         logger,

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -442,7 +442,7 @@ func (c *EpinioClient) Services() error {
 		WithStringValue("Organization", c.Config.Org).
 		Msg("Listing services")
 
-	details.Info("list applications")
+	details.Info("list services")
 
 	jsonResponse, err := c.get(api.Routes.Path("Services", c.Config.Org))
 	if err != nil {
@@ -477,17 +477,20 @@ func (c *EpinioClient) ServiceMatching(ctx context.Context, prefix string) []str
 
 	result := []string{}
 
-	// TODO Create and use server endpoints. Maybe use existing
-	// `Index`/Listing endpoint, either with parameter for
-	// matching, or local matching.
+	// Ask for all services. Filtering is local.
+	// TODO: Create new endpoint (compare `EnvMatch`) and move filtering to the server.
 
-	orgServices, err := services.List(ctx, c.Cluster, c.Config.Org)
+	jsonResponse, err := c.get(api.Routes.Path("Services", c.Config.Org))
 	if err != nil {
 		return result
 	}
+	var response models.ServiceResponseList
+	if err := json.Unmarshal(jsonResponse, &response); err != nil {
+		return result
+	}
 
-	for _, s := range orgServices {
-		service := s.Name()
+	for _, s := range response {
+		service := s.Name
 		details.Info("Found", "Name", service)
 		if strings.HasPrefix(service, prefix) {
 			details.Info("Matched", "Name", service)

--- a/internal/cli/clients/client.go
+++ b/internal/cli/clients/client.go
@@ -806,27 +806,30 @@ func (c *EpinioClient) Info() error {
 	log.Info("start")
 	defer log.Info("return")
 
-	platform := c.Cluster.GetPlatform()
-	kubeVersion, err := c.Cluster.GetVersion()
-	if err != nil {
-		return errors.Wrap(err, "failed to get kube version")
-	}
-
 	// TODO: Extend the epinio API to get the gitea version
 	// information again. Or remove it entirely.
+	// Note: See also Spike #699 for possible full removal of Gitea
 
 	giteaVersion := "unavailable"
-
 	epinioVersion := "unavailable"
+	platform := "unknown"
+	kubeVersion := "unknown"
+
 	if jsonResponse, err := c.get(api.Routes.Path("Info")); err == nil {
-		v := struct{ Version string }{}
+		v := struct {
+			Version     string
+			KubeVersion string
+			Platform    string
+		}{}
 		if err := json.Unmarshal(jsonResponse, &v); err == nil {
 			epinioVersion = v.Version
+			platform = v.Platform
+			kubeVersion = v.KubeVersion
 		}
 	}
 
 	c.ui.Success().
-		WithStringValue("Platform", platform.String()).
+		WithStringValue("Platform", platform).
 		WithStringValue("Kubernetes Version", kubeVersion).
 		WithStringValue("Gitea Version", giteaVersion).
 		WithStringValue("Epinio Version", epinioVersion).

--- a/internal/cli/clients/push.go
+++ b/internal/cli/clients/push.go
@@ -114,7 +114,7 @@ func (c *EpinioClient) stageLogs(ctx context.Context, details logr.Logger, appRe
 	return err
 }
 
-func (c *EpinioClient) deployCode(req models.DeployRequest) ([]byte, error) {
+func (c *EpinioClient) deployCode(req models.DeployRequest) (*models.DeployResponse, error) {
 	out, err := json.Marshal(req)
 	if err != nil {
 		return nil, errors.Wrap(err, "can't marshal deploy request")
@@ -125,7 +125,13 @@ func (c *EpinioClient) deployCode(req models.DeployRequest) ([]byte, error) {
 		return nil, errors.Wrap(err, "can't deploy app")
 	}
 
-	return b, nil
+	// returns app default route
+	deploy := &models.DeployResponse{}
+	if err := json.Unmarshal(b, deploy); err != nil {
+		return nil, err
+	}
+
+	return deploy, nil
 }
 
 func (c *EpinioClient) waitForPipelineRun(ctx context.Context, app models.AppRef, id string) error {

--- a/internal/cli/clients/push.go
+++ b/internal/cli/clients/push.go
@@ -7,18 +7,12 @@ import (
 	"io/ioutil"
 	"path"
 	"sync"
-	"time"
 
-	"github.com/epinio/epinio/deployments"
 	api "github.com/epinio/epinio/internal/api/v1"
 	"github.com/epinio/epinio/internal/api/v1/models"
-	"github.com/epinio/epinio/internal/duration"
 	"github.com/go-logr/logr"
 	"github.com/mholt/archiver/v3"
 	"github.com/pkg/errors"
-	tekton "github.com/tektoncd/pipeline/pkg/client/clientset/versioned"
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/util/wait"
 )
 
 func collectSources(log logr.Logger, source string) (string, string, error) {
@@ -137,47 +131,13 @@ func (c *EpinioClient) deployCode(req models.DeployRequest) ([]byte, error) {
 func (c *EpinioClient) waitForPipelineRun(ctx context.Context, app models.AppRef, id string) error {
 	c.ui.ProgressNote().KeeplineUnder(1).Msg("Running staging")
 
-	cs, err := tekton.NewForConfig(c.Cluster.RestConfig)
-	if err != nil {
-		return err
-	}
-	client := cs.TektonV1beta1().PipelineRuns(deployments.TektonStagingNamespace)
-
-	return wait.PollImmediate(time.Second, duration.ToAppBuilt(),
-		func() (bool, error) {
-			l, err := client.List(ctx, metav1.ListOptions{LabelSelector: models.EpinioStageIDLabel + "=" + id})
-			if err != nil {
-				return false, err
-			}
-			if len(l.Items) == 0 {
-				return false, nil
-			}
-			for _, pr := range l.Items {
-				// any failed conditions, throw an error so we can exit early
-				for _, c := range pr.Status.Conditions {
-					if c.IsFalse() {
-						return false, errors.New(c.Message)
-					}
-				}
-				// it worked
-				if pr.Status.CompletionTime != nil {
-					return true, nil
-				}
-			}
-			// pr exists, but still running
-			return false, nil
-		})
+	_, err := c.get(api.Routes.Path("StagingComplete", app.Org, id))
+	return err
 }
 
 func (c *EpinioClient) waitForApp(ctx context.Context, app models.AppRef) error {
 	c.ui.ProgressNote().KeeplineUnder(1).Msg("Creating application resources")
 
-	err := c.Cluster.WaitForDeploymentCompleted(
-		ctx,
-		c.ui, app.Org, app.Name, duration.ToAppBuilt())
-	if err != nil {
-		return errors.Wrap(err, "waiting for app to come online failed")
-	}
-
-	return nil
+	_, err := c.get(api.Routes.Path("AppRunning", app.Org, app.Name))
+	return err
 }

--- a/internal/domain/domain.go
+++ b/internal/domain/domain.go
@@ -17,7 +17,7 @@ func MainDomain(ctx context.Context) (string, error) {
 		return mainDomain, nil
 	}
 
-	cluster, err := kubernetes.GetCluster(ctx)
+	cluster, err := kubernetes.GetCluster(ctx) // Push -> MainDomain -> here (client-side! FIXIT?)
 	if err != nil {
 		return "", err
 	}

--- a/internal/domain/domain.go
+++ b/internal/domain/domain.go
@@ -17,7 +17,7 @@ func MainDomain(ctx context.Context) (string, error) {
 		return mainDomain, nil
 	}
 
-	cluster, err := kubernetes.GetCluster(ctx) // Push -> MainDomain -> here (client-side! FIXIT?)
+	cluster, err := kubernetes.GetCluster(ctx)
 	if err != nil {
 		return "", err
 	}

--- a/internal/domain/domain.go
+++ b/internal/domain/domain.go
@@ -2,6 +2,7 @@ package domain
 
 import (
 	"context"
+	"fmt"
 	"strings"
 
 	"github.com/epinio/epinio/deployments"
@@ -10,6 +11,14 @@ import (
 )
 
 var mainDomain = ""
+
+func AppDefaultRoute(ctx context.Context, name string) (string, error) {
+	mainDomain, err := MainDomain(ctx)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s.%s", name, mainDomain), nil
+}
 
 // GetMain finds the main domain for Epinio
 func MainDomain(ctx context.Context) (string, error) {

--- a/internal/web/info_controller.go
+++ b/internal/web/info_controller.go
@@ -3,7 +3,7 @@ package web
 import (
 	"net/http"
 
-	"github.com/epinio/epinio/internal/cli/clients"
+	"github.com/epinio/epinio/helpers/kubernetes"
 	"github.com/epinio/epinio/internal/cli/clients/gitea"
 	"github.com/epinio/epinio/internal/version"
 )
@@ -13,18 +13,19 @@ type InfoController struct {
 
 func (hc InfoController) Index(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
-	client, err := clients.NewEpinioClient(ctx)
-	if handleError(w, err, 500) {
-		return
-	}
 
 	giteaClient, err := gitea.New(ctx)
 	if handleError(w, err, 500) {
 		return
 	}
 
-	platform := client.Cluster.GetPlatform()
-	kubeVersion, err := client.Cluster.GetVersion()
+	cluster, err := kubernetes.GetCluster(ctx)
+	if handleError(w, err, 500) {
+		return
+	}
+
+	platform := cluster.GetPlatform()
+	kubeVersion, err := cluster.GetVersion()
 	if handleError(w, err, 500) {
 		return
 	}

